crates/red_knot_python_semantic/src/types.rs

```diff
@@ -4204,6 +4211,18 @@ impl<'db> UnionType<'db> {
     ) -> Type<'db> {
         Self::from_elements(db, self.elements(db).iter().map(transform_fn))
     }
+
+    #[salsa::tracked]
```

Comment:
Did you measure the performance impact of caching the sorted unions? 

I'm asking because it's unclear to me if caching it always is worth it. E.g. sorting a union with only a few elements is super cheap. So maybe it's only worth calling into the cached variant after a certain threshold? 

It might also be worth to test first if the union elements aren't sorted already

Reply:
I pushed a commit shortly before your review to see what happens to the benchmarks if we don't cache them. Though I'm still worried about what will happen if we encounter very large unions, which do occur regularly in popular Python libraries such as pydantic (see https://github.com/astral-sh/ruff/issues/13549).

Reply:
The benchmarks do seem to be much improved without the caching (codspeed is [now neutral](https://codspeed.io/astral-sh/ruff/branches/alex%2Funion-equivalence)). So I guess it makes sense to keep these uncached for now until we benchmark how red-knot performs on projects with catastrophically huge unions.

Reply:
> It might also be worth to test first if the union elements aren't sorted already

it [would be easier](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.is_sorted_by) if we upgraded our MSRV to 1.82 :P

Reply:
It shouldn't be too hard to implement ;)

Reply:
done in https://github.com/astral-sh/ruff/pull/15516/commits/1e4769339fb2e54eff016ab304f2361aaaa69528 :-)

Reply:
The version where we first check to see whether it is already sorted before attempting to sort it does (I think) seem to be doing _slightly_ worse on the benchmarks? But it's all within the noise range; I might be looking for signal here in something that's actually random variation.

Reply:
I just noticed that checking `is_sorted` first is probably wasteful because rustc hash function aims to be `O(n)` when the import is already sorted. 

I also think that we should probably leave those optimizations for later and only apply them based on the size of the union (in which case we may also decide to cache). 

Sorry that I misguided you here.